File: next-macos-native/Sources/MooToolNextCore/LegacyTimeConvertDraft.swift
import Foundation

enum LegacyTimeConvertDraft {
    struct ParsedEntry: Equatable, Sendable {
        var summary: String
        var input: String
        var output: String
        var zone: String
        var unit: String
    }

    private static let timestampToLocal = try! NSRegularExpression(
        pattern: #"^(?:时间戳|Timestamp|タイムスタンプ):\s*(\d+)\s*-->\s*(?:时间|Time|時間)\(([^)]+)\):\s*(.+)$"#,
        options: [.caseInsensitive])
    private static let localToTimestamp = try! NSRegularExpression(
        pattern: #"^(?:时间|Time|時間)\s*\(([^)]+)\):\s*(.+?)\s*-->\s*(?:时间戳|Timestamp|タイムスタンプ):\s*(\d+)$"#,
        options: [.caseInsensitive])

    static func parse(_ content: String) -> [ParsedEntry] {
        let fromConsole = LegacyConsoleDraft.extractMessages(content).compactMap { parseLine($0.trimmingCharacters(in: .whitespacesAndNewlines)) }
        if !fromConsole.isEmpty { return fromConsole }
        return content.split(whereSeparator: \.isNewline)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .compactMap(parseLine)
    }

    static func parseLine(_ line: String) -> ParsedEntry? {
        let range = NSRange(line.startIndex..., in: line)
        if let match = timestampToLocal.firstMatch(in: line, range: range) {
            let timestamp = capture(line, match: match, index: 1)
            let zone = capture(line, match: match, index: 2).trimmingCharacters(in: .whitespacesAndNewlines)
            let localTime = capture(line, match: match, index: 3).trimmingCharacters(in: .whitespacesAndNewlines)
            return ParsedEntry(summary: line, input: timestamp, output: localTime, zone: zone, unit: unitForTimestamp(timestamp))
        }
        if let match = localToTimestamp.firstMatch(in: line, range: range) {
            let zone = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            let localTime = capture(line, match: match, index: 2).trimmingCharacters(in: .whitespacesAndNewlines)
            let timestamp = capture(line, match: match, index: 3)
            return ParsedEntry(summary: line, input: localTime, output: timestamp, zone: zone, unit: unitForTimestamp(timestamp))
        }
        return nil
    }

    private static func capture(_ line: String, match: NSTextCheckingResult, index: Int) -> String {
        guard match.numberOfRanges > index, let range = Range(match.range(at: index), in: line) else { return "" }
        return String(line[range])
    }

    private static func unitForTimestamp(_ timestamp: String) -> String {
        timestamp.count >= 13 ? "millisecond" : "second"
    }
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/domain/LegacyTimeConvertDraft.kt
package com.rememberber.mootool.next.compose.domain

/**
 * Parses Java MooTool `TimeConvert` left-panel log (`t_func_content`) into compose time-convert history rows.
 */
object LegacyTimeConvertDraft {
    private val timestampToLocal = Regex(
        """^(?:时间戳|Timestamp|タイムスタンプ):\s*(\d+)\s*-->\s*(?:时间|Time|時間)\(([^)]+)\):\s*(.+)$""",
        RegexOption.IGNORE_CASE
    )
    private val localToTimestamp = Regex(
        """^(?:时间|Time|時間)\s*\(([^)]+)\):\s*(.+?)\s*-->\s*(?:时间戳|Timestamp|タイムスタンプ):\s*(\d+)$""",
        RegexOption.IGNORE_CASE
    )

    data class ParsedEntry(
        val summary: String,
        val input: String,
        val output: String,
        val zone: String,
        val unit: String,
    )

    fun parse(content: String): List<ParsedEntry> {
        val fromConsole = LegacyConsoleDraft.extractMessages(content)
            .mapNotNull { message -> parseLine(message.trim()) }
        if (fromConsole.isNotEmpty()) return fromConsole
        return content.lineSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .mapNotNull { line -> parseLine(line) }
            .toList()
    }

    fun parseLine(line: String): ParsedEntry? {
        timestampToLocal.matchEntire(line)?.let { match ->
            val timestamp = match.groupValues[1]
            val zone = match.groupValues[2].trim()
            val localTime = match.groupValues[3].trim()
            return ParsedEntry(
                summary = line,
                input = timestamp,
                output = localTime,
                zone = zone,
                unit = unitForTimestamp(timestamp),
            )
        }
        localToTimestamp.matchEntire(line)?.let { match ->
            val zone = match.groupValues[1].trim()
            val localTime = match.groupValues[2].trim()
            val timestamp = match.groupValues[3]
            return ParsedEntry(
                summary = line,
                input = localTime,
                output = timestamp,
                zone = zone,
                unit = unitForTimestamp(timestamp),
            )
        }
        return null
    }

    fun optionsJson(zone: String, unit: String): String {
        val unitEnum = if (unit == "millisecond") TimestampUnit.Millisecond else TimestampUnit.Second
        return TimeHistoryMetadata.encode(zone, unitEnum)
    }

    private fun unitForTimestamp(timestamp: String): String =
        if (timestamp.length >= 13) "millisecond" else "second"
}
